File: challenges/leet_code/2958/solution.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int MaxSubarrayLength(int[] nums, int k) {
                int left = 0;
        Dictionary<int, int> counter = new Dictionary<int, int>();
        int max_length = 0;
        for (int right = 0; right < nums.Length; right++) {
            if (counter.ContainsKey(nums[right])) {
                counter[nums[right]] += 1;
            } else {
                counter[nums[right]] = 1;
            }
            while (counter[nums[right]] > k) {
                counter[nums[left]] -= 1;
                if (counter[nums[left]] == 0) {
                    counter.Remove(nums[left]);
                }
                left += 1;
            }
            max_length = Math.Max(max_length, right - left + 1);
        }
        return max_length;
    }
}

File: challenges/leet_code/2958/solution.cpp
#include <unordered_map>
#include <algorithm>


class Solution {
    public:
        int maxSubarrayLength(vector<int>& nums, int k) {
            int left = 0;
    std::unordered_map<int, int> counter;
    int max_length = 0;
    for (int right = 0; right < nums.size(); ++right) {
        ++counter[nums[right]];
        while (counter[nums[right]] > k) {
            --counter[nums[left]];
            if (counter[nums[left]] == 0) {
                counter.erase(nums[left]);
            }
            ++left;
        }
        max_length = std::max(max_length, right - left + 1);
    }
    return max_length;

        }
};
